### ISSUE TITLE
There is no documentation on the shape of the dashboard string

### ISSUE THREAD

**Issue by bashbreakpoint** [newrelic/newrelic-cloudformation-resource-providers-dashboards#1]
I have been trying to use this plugin for a couple of days now, knowing it's experimental, and I get no dashboard and poor feedback. My issues are below. I'm not expecting a fix right away, and I would be happy to help with documentation to get the next person over the hump.
1. I have used both the `@cdk-cloudformation/newrelic-observability-dashboards` and `@cdk-cloudformation/newrelic-cloudformation-dashboards` connected with the appropriately activated plugins, with the Cloudformation Dashbord plugin configured properly at the account level.
2. On all of my attempts to deploy, it runs for 30 minutes or so before a rollback with the only error message being `Internal Error`.
3. I have tried several shapes of dashboard json string. Initially, I went with a shape very much like that in the CDKTF provider (built from the New Relic Terraform Provider). It looks much like the NerdGraph call.
4. I built a query in NerdGraph and used that shape for a dashboard.
```
      const buildADashboard = {
        dashboard: {
          description: 'Nerdgraph Test Dashboard',
          name: 'Nerdgraph Test Dashboard',
          pages: {
            description: 'Page',
            name: 'Page',
            widgets: { title: 'Widget Title', configuration: { markdown: { text: 'Some Markdown' } } },
          },
          permissions: 'PUBLIC_READ_ONLY',
        },
      }
      
     new CFNDashboards(this, 'DashboardThing', {
        dashboard: JSON.stringify(buildADashboard),
      })
    }
```
is the minimum example I have (in CDK). The shape works through a direct graphql call in the NerdGraph explorer, but not in CDK.
5. I can't find any documentation on what's expected in that string other than pointers to NerdGraph.
6. A minimum working example would be lovely. If we can get one working, I would be happy to PR it back to the README here.
7. Better error messaging other than "Internal Error" would go a long way to help users self-solve problems. I'd also be willing to help with that, though my `Go` skills are nearly non-existent.

**Comment by msummers-nr** [comment]
@bashbreakpoint All of our work has been in the CF Console and  the AWS CLI, we don't have any experience with CDK.

If the `dashboard: ` works directly in NerdGraph then it sounds like the `dashboard: ` string literal is getting mangled, do you have any way of generating debug output for CDK? A larger snippet of the CDK script might show something, it would be interesting to know what `CFNDashboards` is doing. 

I'm wondering if what's sent to the custom resource starts with `{ dashboard: ` due to `buildDashboard` being an object rather than a string.

Under `Description` in the `README.md` there is a link to the NerdGraph Dashboard API docs. NerdGraph is the final word however for the shape of the `dashboard:` string.

**Comment by bashbreakpoint** [comment]
This is all good information, thank you for the time. Noticing the `dashboard` was an object seems to be a good pointer. I'll experiment more with that. I'm thinking it probably was adding an extra `{}` wrapper around the construct and that could be part of the problem. Here is my full current CDK example that I'm working on:
```
import { Stack, StackProps } from 'aws-cdk-lib'
import { Construct } from 'constructs'
import { CfnDashboards as O11yDashboard } from '@cdk-cloudformation/newrelic-observability-dashboards'

export class DashboardHackStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props)

    const buildADashboard = {
      description: 'Nerdgraph Test Dashboard',
      name: 'Nerdgraph Test Dashboard',
      pages: {
        description: 'Page',
        name: 'Page',
        widgets: { title: 'Widget Title', configuration: { markdown: { text: 'Some Markdown' } } },
      },
      permissions: 'PUBLIC_READ_ONLY',
    }

    new O11yDashboard(this, 'DashboardThing', {
      dashboard: `dashboard: ${JSON.stringify(buildADashboard)}`,
    })
  }
}
```
and this synthesizes down into CloudFormation JSON as thus:
```
{
 "Resources": {
  "DashboardThing": {
   "Type": "NewRelic::Observability::Dashboards",
   "Properties": {
    "Dashboard": "dashboard: {\"description\":\"Nerdgraph Test Dashboard\",\"name\":\"Nerdgraph Test Dashboard\",\"pages\":{\"description\":\"Page\",\"name\":\"Page\",\"widgets\":{\"title\":\"Widget Title\",\"configuration\":{\"markdown\":{\"text\":\"Some Markdown\"}}}},\"permissions\":\"PUBLIC_READ_ONLY\"}"
   },
   "Metadata": {
    "aws:cdk:path": "litton-oac-experiment-dev/DashboardThing"
   }
  },
  "CDKMetadata": {
   "Type": "AWS::CDK::Metadata",
   "Properties": {
    "Analytics": "v2:deflate64:H4sIAAAAAAAA/zPSM7fUM1RMLC/WTU7J1s3JTNKrDi5JTM7WcU7LC0otzi8tSk6t1cnLT0nVyyrWLzO00DM01TNQzCrOzNQtKs0rycxN1QuC0ACYlYR7TAAAAA=="
   },
   "Metadata": {
    "aws:cdk:path": "litton-oac-experiment-dev/CDKMetadata/Default"
   }
  }
 },
 "Parameters": {
  "BootstrapVersion": {
   "Type": "AWS::SSM::Parameter::Value<String>",
   "Default": "/cdk-bootstrap/hnb659fds/version",
   "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"
  }
 },
 "Rules": {
  "CheckBootstrapVersion": {
   "Assertions": [
    {
     "Assert": {
      "Fn::Not": [
       {
        "Fn::Contains": [
         [
          "1",
          "2",
          "3",
          "4",
          "5"
         ],
         {
          "Ref": "BootstrapVersion"
         }
        ]
       }
      ]
     },
     "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
    }
   ]
  }
 }
}
```

This looks not to be working, but it takes 30 minutes to totally fail. I will report back later today. I suspect, if it's not working, the fix should likely be removing the extra `dashboard` layer I put in. In retrospect, that was a silly mistake. I'll try later today without that part of it and see what happens. I'll report back with the CDK and synthed CF json in either case.

Thank you again for your time. And as I said above, once we have a working example, I'm happy to push it back into the README.

**Comment by msummers-nr** [comment]
Thanks for the update. 

Rather than add to the README let's add an `examples` directory and put your working CDK there as `cdk.ts`

Keep me posted, thanks again for the update.

**Comment by bashbreakpoint** [comment]
I removed the top-level `dashboard` segment and it's still a no-go. The synthed Cloudformation is
```
{
 "Resources": {
  "DashboardThing": {
   "Type": "NewRelic::Observability::Dashboards",
   "Properties": {
    "Dashboard": "{\"description\":\"Nerdgraph Test Dashboard\",\"name\":\"Nerdgraph Test Dashboard\",\"pages\":{\"description\":\"Page\",\"name\":\"Page\",\"widgets\":{\"title\":\"Widget Title\",\"configuration\":{\"markdown\":{\"text\":\"Some Markdown\"}}}},\"permissions\":\"PUBLIC_READ_ONLY\"}"
   },
   "Metadata": {
    "aws:cdk:path": "litton-oac-experiment-dev/DashboardThing"
   }
  },
  "CDKMetadata": {
   "Type": "AWS::CDK::Metadata",
   "Properties": {
    "Analytics": "v2:deflate64:H4sIAAAAAAAA/zPSM7fUM1RMLC/WTU7J1s3JTNKrDi5JTM7WcU7LC0otzi8tSk6t1cnLT0nVyyrWLzO00DM01TNQzCrOzNQtKs0rycxN1QuC0ACYlYR7TAAAAA=="
   },
   "Metadata": {
    "aws:cdk:path": "litton-oac-experiment-dev/CDKMetadata/Default"
   }
  }
 },
 "Parameters": {
  "BootstrapVersion": {
   "Type": "AWS::SSM::Parameter::Value<String>",
   "Default": "/cdk-bootstrap/hnb659fds/version",
   "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"
  }
 },
 "Rules": {
  "CheckBootstrapVersion": {
   "Assertions": [
    {
     "Assert": {
      "Fn::Not": [
       {
        "Fn::Contains": [
         [
          "1",
          "2",
          "3",
          "4",
          "5"
         ],
         {
          "Ref": "BootstrapVersion"
         }
        ]
       }
      ]
     },
     "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
    }
   ]
  }
 }
}
```
Looking at the json, do you have insight from the inside what the dashboard object _should_ look like?

**Comment by msummers-nr** [comment]
Yes, it should look like
```
    "Dashboard": "dashboard: {\"description\":\"Nerdgraph Test Dashboard\",\"name\":\"Nerdgraph Test Dashboard\",\"pages\":{\"description\":\"Page\",\"name\":\"Page\",\"widgets\":{\"title\":\"Widget Title\",\"configuration\":{\"markdown\":{\"text\":\"Some Markdown\"}}}},\"permissions\":\"PUBLIC_READ_ONLY\"}"

```
Look at [inputs_1_create](https://github.com/newrelic-experimental/newrelic-cloudformation-resource-providers-dashboards/blob/main/inputs/inputs_1_create.json), that works.

I'd go with your _original_ js object and then remove the outer most curly brace pair _after_ stringifying.

**Comment by bashbreakpoint** [comment]
Update, but not working yet. I diffed the string that came out of the inputs_1_create and what I was doing. It appears to me that I need to strip quotes off the keys. Tried that. If I look at json-ified output to compare resources, I have an exact match on that string (after I remove some spaces and stuff). It's still not deploying, but I think I've at least got the string part now. I've got a weird week, so it might be a bit until my next update.

**Comment by msummers-nr** [comment]
@bashbreakpoint  We may be running into an issue with the differences between GraphQL and JSON, they're not quite compatible and the `dashbard: ` string is GraphQL rather than JSON.  Something like this (untested) should help
```
JSON.stringify(buildADashboard).replace(/"([^"]+)":/g, '$1:')
```
as GraphQL keys are not quoted.

**Comment by bashbreakpoint** [comment]
Yeah. I figured that out diffing a json-ified version of the yaml file and what cdk produced. I ended up needing
```
  function toGqlString(obj: any): string {
    const firstPass = JSON.stringify(obj).replace(/\"([\w_-]+?)\"\:/g, '$1:')
    const permissionsUpdated = firstPass.replace(/permissions:\"([A-Z]+)\"/g, 'permissions:$1')
    return permissionsUpdated
  }
  ```
  because permissions is also not quoted. I'm at the point now that I have an exact replica of the string, but still no luck. I can `curl` (well, `httpie`) to the NerdGraph endpoint with the string produced and it will actually create the dashboard. But when I run it through CloudFormation, I'm still not getting a dashboard created and the only feedback is `Internal Error`. I'm double checking all my config for the plugin, and it still looks right to me. It's this:
```
{
"APIKey":"NRAK-<redacted>",
"AccountID":"<redacted>",
"Endpoint": "https://api.newrelic.com/graphql"
}
```
